File: erizo/src/erizo/handlers/parseHandlers.py
#!/usr/bin/env python
import json
import ast


from handlers import handlers 

map = "{"
handlerEnum = "{"

if (len(handlers)>0):
	for handler in handlers:
		handlerName = handler["handlerName"]
		handlerClassName = handler["className"]
		map += ("{\"" + handlerName + "\" , " + handlerName + "Enum}," )
		handlerEnum += (handlerName + "Enum,")
	map = map[0:-1] + "}"
	handlerEnum = handlerEnum[0:-1] + "}"
else: 
	map ="{}"
	handlerEnum = "{}"



file = open("HandlerImporter.h", "r") 
lines = file.readlines();
linesOutput = []

for line in lines:
	if (line.find("*handlerEnum*") != -1):
		linesOutput.append("\tenum HandlersEnum "+ handlerEnum + ";\n")
	elif (line.find("*handlerMap*") != -1):
		linesOutput.append("\tstd::map<std::string, HandlersEnum> handlersDic =" + map + ";\n")
	elif (line.find("*imports*") != -1):
		for handler in handlers:
			handlerClassName = handler["className"]
			linesOutput.append("#include \"" + handlerClassName + ".h\"\n")
	else:
		linesOutput.append(line)


file = open("HandlerImporter.h", "w")
lines = file.writelines(linesOutput);
file.close()


file = open("HandlerImporter.cpp", "r") 
lines = file.readlines()
linesOutput = []

for line in lines:
	if (line.find("*case*") != -1):
		for handler in handlers:
			handlerName = handler["handlerName"]
			handlerClassName = handler["className"]
			handlerEnum = (handlerName + "Enum")
			linesOutput.append("\t\tcase "+ handlerEnum +":\n")
			linesOutput.append("\t\t\tptr = std::make_shared<"+handlerClassName+">(parameters);\n")
			linesOutput.append("\t\t\tbreak;\n")

	else:
		linesOutput.append(line);
		
file = open("HandlerImporter.cpp", "w")
lines = file.writelines(linesOutput);
file.close()

File: erizo/src/erizo/handlers/parseHandlers.js
const config = require('../../../../licode_config');

const fs = require('fs');

const handlers = config.erizo.handlers;

let map = '{';
let handlerEnum = '{';

if (handlers.length > 0) {
  handlers.forEach((handler) => {
    const handlerName = handler.handlerName;
    map += `{"${handlerName}", ${handlerName}Enum}, `;
    handlerEnum = `${handlerEnum}${handlerName}Enum, `;
  });
  map = `${map.slice(0, -2)}}`;
  handlerEnum = `${handlerEnum.slice(0, -2)}}`;
} else {
  map = {};
  handlerEnum = {};
}

const headers = fs.readFileSync('HandlerImporter.h', 'utf-8');
let modifiedData = headers.replace('*handlerEnum*', `enum HandlersEnum ${handlerEnum};\n`);
modifiedData = modifiedData.replace('*handlerMap*', `std::map<std::string, HandlersEnum> handlersDic =${map};\n`);

let imports = '';

handlers.forEach((handler) => {
  const handlerClassName = handler.className;
  imports = `${imports}#include "${handlerClassName}.h"\n`;
});
modifiedData = modifiedData.replace('*imports*', imports);

fs.writeFileSync('HandlerImporter.h', modifiedData, 'utf-8');

const cpp = fs.readFileSync('HandlerImporter.cpp', 'utf-8');
let cases = '';

handlers.forEach((handler,index) => {
  const handlerClassName = handler.className;
  const handlerName = handler.handlerName;
  const handlerEnumer = `${handlerName}Enum`;
  if (index === 0) {
    cases = `${cases}case ${handlerEnumer} :\n \t\t\tptr = std::make_shared<${handlerClassName}>(parameters);\n\t\t\tbreak;\n`;
  } else {
    cases = `${cases}\t\tcase ${handlerEnumer} :\n \t\t\tptr = std::make_shared<${handlerClassName}>(parameters);\n\t\t\tbreak;\n`;
  }
});

const modifiedHeaders = cpp.replace('*case*', cases);

fs.writeFileSync('HandlerImporter.cpp', modifiedHeaders, 'utf-8');
